autopilot/hostfilter.go

```diff
@@ -0,0 +1,152 @@
+		// save original GFU & GFR values
+		gfu: m.GoodForUpload,
+		gfr: m.GoodForRenew,
+	}
+}
+
+func (f *hostFilter) finalize() (bus.ContractMetadata, []string, bool) {
+	filtered := (!f.metadata.GoodForUpload && f.gfu) || (!f.metadata.GoodForRenew && f.gfr)
+	return f.metadata, f.reasons, filtered
+}
+
+func (f *hostFilter) withBlackListFilter(cfg Config) *hostFilter {
+	for _, host := range cfg.Hosts.Blacklist {
+		if f.host.IsHost(host) {
+			f.metadata.GoodForUpload = false
+			f.metadata.GoodForRenew = false
+			f.reasons = append(f.reasons, fmt.Sprintf("host is blacklisted, GFU: %v -> false, GFR: %v -> false", f.gfu, f.gfr))
+			break
+		}
+	}
+	return f
+}
+
+func (f *hostFilter) withMaxRevisionFilter() *hostFilter {
+	if f.host.Revision.NewRevisionNumber == math.MaxUint64 {
+		f.metadata.GoodForUpload = false
+		f.metadata.GoodForRenew = false
+		f.reasons = append(f.reasons, fmt.Sprintf("max revision check failed, GFU: %v -> false, GFR: %v -> false", f.gfu, f.gfr))
+	}
+	return f
+}
+
+func (f *hostFilter) withOfflineFilter() *hostFilter {
+	if !f.host.IsOnline() {
+		f.metadata.GoodForUpload = false
+		f.metadata.GoodForRenew = false
+		f.reasons = append(f.reasons, fmt.Sprintf("host is not online, GFU: %v -> false, GFR: %v -> false", f.gfu, f.gfr))
+	}
+	return f
+}
+
+func (f *hostFilter) withRedundantIPFilter(filter *ipFilter) *hostFilter {
+	ip, err := net.ResolveIPAddr("ip", f.host.NetAddress())
+	if err != nil {
+		return f
+	}
+
+	if filter.exists(ip) {
+		f.metadata.GoodForUpload = false
+		f.metadata.GoodForRenew = false
+		f.reasons = append(f.reasons, fmt.Sprintf("host IP is redundant, GFU: %v -> false, GFR: %v -> false", f.gfu, f.gfr))
+	}
+	return f
+}
+
+func (f *hostFilter) withRemainingFundsFilter(cfg Config) *hostFilter {
+	settings, _, found := f.host.LastKnownSettings()
+	if !found {
+		return f
+	}
+
+	blockBytes := types.NewCurrency64(modules.SectorSize * cfg.Contracts.Period)
+	sectorStoragePrice := settings.StoragePrice.Mul(blockBytes)
+	sectorUploadBandwidthPrice := settings.UploadBandwidthPrice.Mul64(modules.SectorSize)
+	sectorDownloadBandwidthPrice := settings.DownloadBandwidthPrice.Mul64(modules.SectorSize)
+	sectorBandwidthPrice := sectorUploadBandwidthPrice.Add(sectorDownloadBandwidthPrice)
+	sectorPrice := sectorStoragePrice.Add(sectorBandwidthPrice)
+	percentRemaining, _ := big.NewRat(0, 1).SetFrac(f.host.RenterFunds().Big(), f.metadata.TotalCost.Big()).Float64()
+
+	if f.host.RenterFunds().Cmp(sectorPrice.Mul64(3)) < 0 || percentRemaining < minContractFundUploadThreshold {
+		f.metadata.GoodForUpload = false
+		f.reasons = append(f.reasons, fmt.Sprintf("contract has insufficient funds, GFU: %v -> false", f.gfu))
+	}
+	return f
+}
+
+func (f *hostFilter) withScoreFilter(cfg Config, threshold float64) *hostFilter {
+	score := newHostScore(f.host).
+		withAgeScore().
+		withCollateralScore(cfg).
+		withInteractionScore().
+		withSettingsScore(cfg).
+		withUptimeScore().
+		withVersionScore().
+		finalize()
```

Comment:
(nit) might be good to have a helper method which gets the total score of a host. So that we don't end up forgetting to update various places where we want to apply all checks at all times.

Reply:
added in #36